release/src/main/scripts/download_github_actions_artifacts.py

```diff
@@ -0,0 +1,316 @@
+  )
+  start_time = time.time()
+  last_request = start_time
+  spinner = itertools.cycle(["|", "/", "-", "\\"])
+  request_interval = 10
+
+  while True:
+    now = time.time()
+    elapsed_time = time.strftime("%H:%M:%S", time.gmtime(now - start_time))
+    print(
+        f"\r {next(spinner)} Waiting to finish. Elapsed time: {elapsed_time}. "
+        f"Current state: status: `{status}`, conclusion: `{conclusion}`.",
+        end="",
+    )
+
+    time.sleep(0.3)
+    if (now - last_request) > request_interval:
+      last_request = now
+      run_data = get_single_workflow_run_data(run_id, repo_url, github_token)
+      status = safe_get(run_data, "status")
+      conclusion = safe_get(run_data, "conclusion")
+      if status in ["queued", "in_progress"]:
+        continue
+      elif status == "completed" and conclusion == "success":
+        print(
+            f"\rFinished in: {elapsed_time}. "
+            f"Last state: status: `{status}`, conclusion: `{conclusion}`.",
+        )
+        return run_id
+      else:
+        print("\r")
+        raise Exception(
+            f"Run unsuccessful. Conclusion: {conclusion}. Check at: {run_web_url}"
+        )
+
+
+def reset_directory(artifacts_dir):
+  """Clears directory asking for confirmation before."""
+  question = (
+      f"Creating Artifacts directory. Any existing content in {artifacts_dir} will be erased. Proceed?\n"
+      f"Your answer")
+  if get_yes_or_no_answer(question):
+    print(f"Clearing directory: {artifacts_dir}")
+    shutil.rmtree(artifacts_dir, ignore_errors=True)
+    os.makedirs(artifacts_dir)
+  else:
+    print("You said NO for clearing artifacts directory. Quitting ...")
+    sys.exit(1)
+
+
+def download_artifacts(run_id, repo_url, artifacts_dir, github_token):
+  """Downloads github artifacts from given run."""
+  print("Starting downloading artifacts ... (it may take a while)")
+  run_data = get_single_workflow_run_data(run_id, repo_url, github_token)
+  artifacts_url = safe_get(run_data, "artifacts_url")
+  data_artifacts = request_url(artifacts_url, github_token)
+  artifacts = safe_get(data_artifacts, "artifacts", artifacts_url)
+  filtered_artifacts = [
+      a for a in artifacts if (
+          a["name"].startswith("source_gztar_zip") or
+          a["name"].startswith("wheelhouse"))
+  ]
+  for artifact in filtered_artifacts:
+    url = safe_get(artifact, "archive_download_url")
+    name = safe_get(artifact, "name")
+    size_in_bytes = safe_get(artifact, "size_in_bytes")
+
+    download_single_artifact(
+        url, name, size_in_bytes, artifacts_dir, github_token)
+
+
+def download_single_artifact(
+    url, name, size_in_bytes, artifacts_dir, github_token):
+  """Downloads single github artifact."""
+  artifacts_size_mb = round(size_in_bytes / (1024 * 1024), 2)
+  print(
+      f"\tDownloading {name}.zip artifact (size: {artifacts_size_mb} megabytes)"
+  )
+
+  with tempfile.NamedTemporaryFile(
+    "wb", prefix=name, suffix=".zip"
```

Comment:
It's not clear from reading this code why there is an assumption that the 'single artifact' will be a 'zip archive', and why download_single_artifact also extracts the archive.

Reply:
I refactored it. I made `download_single_artifact` and `extract_single_artifact` methods and updated docstring.